src/CodeStyle/Core/Analyzers/AbstractCodeStyleDiagnosticAnalyzer.cs

```diff
@@ -69,6 +57,23 @@ protected DiagnosticDescriptor CreateUnnecessaryDescriptor(string descriptorId)
                 descriptorId, _localizableTitle, _localizableMessageFormat,
                 isUnneccessary: true);
 
+        /// <summary>
+        /// Diagnostic descriptor for code you want to fade out and do *not* want to have a smart-tag
+        /// appear for.  This is uncommon but useful in some cases.  For example, if you are fading
+        /// out pieces of code before/after another piece of code *on the same line*, then you will
+        /// only want one usage of <see cref="UnnecessaryWithSuggestionDescriptor"/> and multiple
+        /// usages of <see cref="CreateUnnecessaryWithoutSuggestionDiagnostic"/>.
+        ///
+        /// That's because if you use <see cref="UnnecessaryWithSuggestionDescriptor"/> for all the
+        /// faded out code then that will mean the user will see multiple code actions to fix the
+        /// same issue when they bring up the code action on that line.  Using these two descriptors
+        /// helps ensure that there will not be useless code-action overload.
+        /// </summary>
+        protected Diagnostic CreateUnnecessaryWithoutSuggestionDiagnostic(Location location)
```

Comment:
@mavasani did we ever add the ability to specify faded regions without additional diagnostics?

Reply:
Yes, @dibarbet added the helper API to create a diagnostic with additional locations for fading also fixed up one of the analyzers. We should clean up all our IDE analyzers to use the same approach instead of using separate fading diagnostics.

Reply:
Would it make sense for me to add that cleanup as a commit to this PR?

Reply:
Would that fix mean removing just `UnnecessaryWithoutSuggestion` or both that and `Unnecessary`?

Reply:
DiagnosticHelper.CreateWithLocationTags itself needs a helper

Reply:
Does https://github.com/dotnet/roslyn/pull/41854/commits/1812d310fd82bf74fd981f75b19ecc8cf4e7111c seem like a good direction to go in? If not, let me know what you're thinking or just take over and I'll close the PR.

Reply:
I think `UnnecessaryWithSuggestion` is still needed, because it has a diagnostic ID that gets picked up by a code fix. `UnnecessaryWithoutSuggestion` exists solely for the fading.

Reply:
@sharwell @masavani The helper API created by @dibarbet does not fade the additional locations unless the diagnostic descriptor used *also* has the 'Unnecessary' tag. This means that the new API can't be used to report a diagnostic on a non-faded span and additional faded spans at the same time.

Plenty of analyzers using `UnnecessaryWithoutSuggestion` are just this kind: they are fading spans *other than* the main reported span which should not be faded, and one analyzer even uses `UnnecessaryWithoutSuggestion` to fade a span specifically without allowing Ctrl+. to register an action for that span.

I'm not sure what the best course of action is, but my first thought is that this new API is not ready to be used in the timeframe for this PR to be merged. Does this sound right to you?

Reply:
@CyrusNajmabadi Before you keep reviewing, I'm going to drop all but the first two commits (the ones that actually fix the titular bug).

@sharwell @mavasani @CyrusNajmabadi I can't do the cleanup mentioned above without direction. See the [explanation](https://github.com/dotnet/roslyn/pull/41854#discussion_r382925338) of the situations I ran into that the new `CreateWithLocationTags` can't handle.